src/google/protobuf/generated_message_reflection.cc

```diff
@@ -1223,6 +1223,38 @@ void Reflection::SetString(Message* message, const FieldDescriptor* field,
 }
 
 
+void Reflection::SetString(Message* message, const FieldDescriptor* field,
+                           const std::string&& value) const {
+  USAGE_CHECK_ALL(SetString, SINGULAR, STRING);
+  if (field->is_extension()) {
+    return MutableExtensionSet(message)->SetString(field->number(),
+                                                   field->type(), value, field);
```

Comment:
Also here i believe the value parameter (i.e. the input string) should be moved, so:
`SetString(...,..., std::move(value), field)`
Additionaly, the string which you are creating in your code should also be moved into place:

`bool foo(const char *buf, size_t size,...) {
  ...
  std::string temp(buf, size); // actually, we don't need to create a temporary std::string here, if we have 
                               // some method like SetStringByArray(..., buf, size); 
  reflection->SetString(...., std::move(temp));
  ...
}`